File: Permutations/Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int> &num) {
        sort(num.begin(), num.end());
        vector<vector<int>> res;
        vector<int> tmp;
        vector<bool> visited(num.size(), false);
        permute(res, tmp, visited, num);
        return res;
    }
    void permute(vector<vector<int>> &res, vector<int> &tmp, vector<bool> visited, vector<int> &num) {
        if (tmp.size() == num.size()) {
            res.push_back(tmp);
        }
        for (int i = 0; i < num.size(); ++i) {
            if (true == visited[i] || (i != 0 && num[i] == num[i-1] && visited[i - 1] == true)) continue;
            tmp.push_back(num[i]);
            visited[i] = true;
            permute(res, tmp, visited, num);
            visited[i] = false;
            tmp.pop_back();
        }
    }
};

File: Permutations/Permutations.java
public class Solution {
    public List<List<Integer>> permute(int[] num) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        Arrays.sort(num);
        permute(res, new ArrayList<Integer>(num.length), new boolean[num.length], num);
        return res;
    }
    private void permute(List<List<Integer>> res, List<Integer> tmp, boolean[] visited, int[] num) {
        if (tmp.size() == num.length) {
            res.add(new ArrayList(tmp));
        }
        for (int i = 0; i < num.length; ++i) {
            if (true == visited[i] || (i != 0 && num[i] == num[i-1] && visited[i - 1] == true)) continue;
            tmp.add(num[i]);
            visited[i] = true;
            permute(res, tmp, visited, num);
            visited[i] = false;
            tmp.remove(tmp.size() - 1);
        }
    }
}